taichi/backends/vulkan/spirv_ir_builder.h

```diff
@@ -0,0 +1,473 @@
+    if (gl_global_invocation_id.id != 0) {
+      ib_.add(gl_global_invocation_id);
+    }
+    if (gl_num_work_groups.id != 0) {
+      ib_.add(gl_num_work_groups);
+    }
+    ib_.commit(&entry_);
+    ib_.begin(spv::OpExecutionMode).add_seq(func, spv::ExecutionModeLocalSize,
+              local_size[0], local_size[1], local_size[2]).commit(&entry_);
+  }
+
+  // Start function scope
+  void start_function(const Value& func) {
+    // add function declaration to the header
+    ib_.begin(spv::OpFunction).add_seq(t_void_, func, 0, t_void_func_).commit(&func_header_);
+
+    spirv::Label start_label = this->new_label();
+    ib_.begin(spv::OpLabel).add_seq(start_label).commit(&func_header_);
+    curr_label_ = start_label;
+  }
+
+  // Declare gl compute shader related methods
+  void set_work_group_size(const std::array<int, 3> group_size);
+  Value get_work_group_size(uint32_t dim_index);
+  Value get_num_work_groups(uint32_t dim_index);
+  Value get_global_invocation_id(uint32_t dim_index);
+
+  // Expressions
+  Value add(Value a, Value b);
+  Value sub(Value a, Value b);
+  Value mul(Value a, Value b);
+  Value div(Value a, Value b);
+  Value mod(Value a, Value b);
+  Value eq(Value a, Value b);
+  Value ne(Value a, Value b);
+  Value lt(Value a, Value b);
+  Value le(Value a, Value b);
+  Value gt(Value a, Value b);
+  Value ge(Value a, Value b);
+  Value select(Value cond, Value a, Value b);
+
+  // Create a cast that cast value to dst_type
+  Value cast(const SType& dst_type, Value value);
+
+  // Create a GLSL450 call
+  template <typename... Args>
+  Value call_glsl450(const SType& ret_type, uint32_t inst_id, Args&&... args) {
+    Value val = new_value(ret_type, ValueKind::kNormal);
+    ib_.begin(spv::OpExtInst).add_seq(ret_type, val, ext_glsl450_, inst_id)
+                             .add_seq(std::forward<Args>(args)...)
+                             .commit(&function_);
+    return val;
+  }
+
+  // Local allocate, load, store methods
+  Value alloca_variable(const SType& type);
+  Value load_variable(Value pointer, const SType& res_type);
+  void store_variable(Value pointer, Value value);
+
+  // Register name to corresponding Value/VariablePointer
+  void register_value(std::string name, Value value);
+  // Query Value/VariablePointer by name
+  Value query_value(std::string name) const;
+
+  // Support easy access to trivial data types
+  SType i32_type() const { return t_int32_; }
+  SType u32_type() const { return t_uint32_; }
+  SType f32_type() const { return t_fp32_; }
+  SType bool_type() const { return t_bool_; }
+  
+  // quick cache for const zero/one i32
+  Value const_i32_zero_;
+  Value const_i32_one_;
+
+  // Use float_atomic_add
+  Value float_atomic_add();
+  Value rand_u32(Value global_tmp_);
+  Value rand_f32(Value global_tmp_);
+  Value rand_i32(Value global_tmp_);
+
+  // Build Snode structure in SPIR-V
+  SType get_snode_struct(const CompiledSNodeStructs* compiled_structs, const SNodeDescriptor& sn_desc);
```

Comment:
nit: These are one level above the SPIR-V builder. Ideally we can have a SNode compiler that uses this IR-builder.

Reply:
Fixed